### DISCUSSION TITLE
Benchmarking against other libraries

### DISCUSSION THREAD

**Discussion by mateuszbaran** [JuliaManifolds/Manopt.jl#338] [General]
I've started this to keep https://github.com/JuliaManifolds/Manopt.jl/issues/333 closed but still discuss benchmarks.

**Comment by mateuszbaran** [comment]
First, Optim.jl and Manopt.jl have slightly different gradient tolerance criteria. Manopt has gradient 2-norm while Optim.jl uses inf-norm: https://julianlsolvers.github.io/Optim.jl/stable/#user/config/

**Reply by kellertuer** [reply]
If I interpret there `g_tol` correctly that should be our https://manoptjl.org/stable/plans/stopping_criteria/#Manopt.StopWhenGradientChangeLess?


Yes the 2-norm is probably more pessimistic, but I am also not so sure how to define an infinity-norm in a tangent space. We currently use the Riemannian metric (which sure is basically the 2-norm), but the infinity-norm would probably require a basis and then one could to the infinity norm of the coefficients with respect to that basis.
One could add that as an option to the criterion above.

**Reply by mateuszbaran** [reply]
Why do you think it's gradient change and not gradient norm?

Inf-norm is indeed basis-dependent but I think we can wave it away as "manifold-dependent"; for stopping criterion that should be good enough.

**Reply by kellertuer** [reply]
For me the text is not completely clear, since it just says “Absolute tolerance in the gradient, in infinity norm” – tolerance to what? The two before speak about change, so from so similar names I concluded it might be gradient change - but nevertheless, also for just the gradient norm, the same problem exists, sure.

What do you mean by “manifold-dependent”? I would say it is basis-dependent, which is a bit more than that. The Riemannian metric would be manifold dependent, but this one is _additionally_ depending on the choice of basis.

However, sure we can add that to either gradient change and/or gradient norm as an optional alternative. The default should stay the one that is easier to use and does not require a basis.

**Reply by sethaxen** [reply]
Agreed, the infinite-norm is basis-dependent. Specifically, I believe the infinite-norm is in the range  `norm(M, p, X) * [inv(sqrt(real_dimension(M))), 1] <= norm(M, p, X)`. So Optim's convergence requirement is easier to satisfy than Manopt's.

**Reply by mateuszbaran** [reply]
> For me the text is not completely clear, since it just says “Absolute tolerance in the gradient, in infinity norm” – tolerance to what? The two before speak about change, so from so similar names I concluded it might be gradient change - but nevertheless, also for just the gradient norm, the same problem exists, sure.

Judging by the results of Optim's state printing, it is gradient norm, at it doesn't even support norm-of-gradient-difference stopping condition.

> What do you mean by “manifold-dependent”? I would say it is basis-dependent, which is a bit more than that. The Riemannian metric would be manifold dependent, but this one is _additionally_ depending on the choice of basis.

Sure, it depends on both. I'm mostly thinking about the possibility of basis being implicit for stopping condition because I don't see much use of inf-norm beyond it being faster to compute and comparing to Optim.jl.

**Reply by kellertuer** [reply]
it is surely possible, though with the basis decomposition we additionally have to do only in rare cases faster.

**Reply by mateuszbaran** [reply]
That's exactly why I'm talking about implicit basis: to avoid basis decomposition.

**Reply by kellertuer** [reply]
This I do not follow. also for an implicit basis I would consider the map that transforms the tangent vector to its coefficients a decomposition. Optim.jl or 2-norm does not do this, so this is usually slow. Sure Euclidean with ONB (=decomp is the identity) there it should be fine ;)

**Reply by mateuszbaran** [reply]
I'm suggesting just calling `norm(X, Inf)` for all tangent vectors represented by arrays :wink:, and adding a mathematical explanation why this is OK. It seems to be. There is *a* basis in which that is an `Inf` norm.

**Comment by kellertuer** [comment]
As a thread this is a great idea to collect and discuss these before opening (maybe more precise issues or) PRs! Thanks!
